### PR TITLE
#241 new package model validation

### PR DESCRIPTION
**Pull request by balusarakesh**
Enforced the correctness of package objects.

### CODE REVIEW

File: src/packagedcode/models.py [hunk 1]
```diff
@@ -117,14 +117,6 @@ class AssertedLicense(Model):
     notice = StringType(default=None)
     url = StringType(default=None)
 
-    def as_dict(self):
```
Comment (pombredanne) [inline src/packagedcode/models.py]: What is this replaced by? the equivalent schematics function?
If this is the case, as_dict is rather clear as a name, I would not want to have the schematics specific to percolat in the codebase everywhere. Instead we should have an as_dict function that wraps schematics calls.

Comment (balusarakesh) [inline src/packagedcode/models.py]: there is an inbuilt function `to_primitive` which can serialize an object...so, I removed `as_dict`

Comment (pombredanne) [inline src/packagedcode/models.py]: ok, so do not remove `as_dict`. Make it available on the top level Package class and just have it call `to_primitive` : the point is that we do not want the external access to Packages to be tied to schematics.

Comment (pombredanne) [inline src/packagedcode/models.py]: Related we should not depend on the default constructor for schematics models to pass a dict vs named arguments: we should instead abstract that and accept **kwargs and pass the kwargs to schematics

Comment (balusarakesh) [inline src/packagedcode/models.py]: @pombredanne  in order to override the default constructor in `schematics` we need to edit the `Model` class in `schematics.models.py`. But everytime we configure `scancode` we get a fresh install of `schematics

Comment (balusarakesh) [inline src/packagedcode/models.py]: maybe we can edit the `schematics` wheel file...is that a good idea?

Comment (pombredanne) [inline src/packagedcode/models.py]: > maybe we can edit the schematics wheel file...is that a good idea?
Not really a good idea.

Comment (pombredanne) [inline src/packagedcode/models.py]: > in order to override the default constructor in schematics we need to edit the Model class in schematics.models.py. But everytime we configure scancode we get a fresh install of schematics and we lose the changes. how can we solve this problem?

do not ever change schematics locally. We can subclass and compose, not override....

File: src/packagedcode/models.py [hunk 2]
```diff
@@ -297,21 +334,21 @@ def compare_version(self, package, package_version=True):
 
         For example:
 
-        >>> q=Package(version='2')
-        >>> p=Package(version='1')
+        >>> q=Package(dict(versioning=Version(dict(version='2'))))
```
Comment (pombredanne) [inline src/packagedcode/models.py]: This is the part we would not want: a direct depdency on the way schematics demands a dict as args.
We want instead to do:
`Package(versioning=Version(version='2'))`
or `Package(versioning=Version('2'))`

File: tests/packagedcode/test_package_models.py [hunk 3]
```diff
@@ -38,7 +41,35 @@ class TestModels(FileBasedTesting):
...
+            vcs_tool='git',
+            asserted_licenses=[
+                AssertedLicense(dict(
+                    license='apache-2.0'
+                    )
+                )
+            ],
+            )
+        )
+        assert 'Sample' == package.name
```
Comment (pombredanne) [inline tests/packagedcode/test_package_models.py]: Could you test the whole result data rather than just some fields?

File: thirdparty/prod/schematics.ABOUT [hunk 4]
```diff
@@ -0,0 +1,11 @@
+about_resource: schematics-1.1.1-cp27-none-any.whl
+version: 1.1.1
+
+name: schematics
+home_url: http://github.com/schematics/schematics
+owner: James Dennis
+dje_license: bsd-new
+notice_file: schematics.NOTICE
```
Comment (pombredanne) [inline thirdparty/prod/schematics.ABOUT]: We need also a `license_text_file: schematics.LICENSE` and the  file and a `copyright:`

File: setup.py [hunk 5]
```diff
@@ -93,6 +93,7 @@ def read(*names, **kwargs):
 
         # packagedcode
         'requests >= 2.7.0, < 3.0.0',
+        'schematics',
```
Comment (pombredanne) [inline setup.py]: You need to put a version range constraint, we do not accept any schematics version but the current up to the next major excluded

File: src/packagedcode/models.py [hunk 6]
```diff
@@ -94,14 +103,80 @@
...
+    party_org = 'organization'
+
+    PARTY_TYPES = (party_person, party_project, party_org,)
+
+    type = StringType(choices=PARTY_TYPES)
+
+
+class Dependency(EnhancedModel):
+    """
+    A dependency points to a Package via a package id and version, and jhas a version
```
Comment (pombredanne) [inline src/packagedcode/models.py]: There is a typo in  `jhas`

File: src/packagedcode/models.py [hunk 7]
```diff
@@ -135,153 +210,121 @@ class Package(object):
     DEPENDENCY_GROUPS = (dep_runtime, dep_dev, dep_optional, dep_test,
                          dep_build, dep_ci, dep_bundled)
 
+    # path to a file or directory
+    location = None
+
+    # the id of the package
+    id = StringType()
```
Comment (pombredanne) [inline src/packagedcode/models.py]: I am not sure we had an id originally? did we? can you refresh me on the purpose?

File: src/packagedcode/models.py [hunk 8]
```diff
@@ -135,153 +210,121 @@ class Package(object):
...
+    versioning = ModelType(Version)
+    name = StringType(required=True)
+    # this is a "short" description.
```
Comment (pombredanne) [inline src/packagedcode/models.py]: Is there a way in schematics to add doc strings to the field directly rather than using comments?

File: src/packagedcode/models.py [hunk 9]
```diff
@@ -135,153 +210,121 @@ class Package(object):
...
+    keywords_doc_url = URLType()
+
+    # paths to metadata files for this package, if any
+    # can be the same as the package location (e.g. RPMs)
+    metafile_locations = ListType(StringType(), default=[])
+
+    # URLs to metadata files for this package.
+    metafile_urls = ListType(URLType(), default=[])
+
+    homepage_url = URLType()
```
Comment (pombredanne) [inline src/packagedcode/models.py]: What are the constraints of a URL type? in some case we have a home that is an FTP url. Does it support this? If not then we should use a plain string instead

File: src/packagedcode/models.py [hunk 10]
```diff
@@ -135,153 +210,121 @@ class Package(object):
...
+    notes = StringType()
+
+    # one or more direct download urls, possibly in SPDX vcs url form
+    # the first one is considered to be the primary
+    download_urls = ListType(URLType(), default=[])
```
Comment (pombredanne) [inline src/packagedcode/models.py]: Same comment as above: we cannot use a URL type if this is limited to  HTTP schemes only.

File: src/packagedcode/models.py (hunk 7 above)
Comment (balusarakesh) [inline src/packagedcode/models.py]: we have an `id` in the original Pacakge class `https://github.com/nexB/scancode-toolkit/blob/develop/src/packagedcode/models.py#L152`

File: src/packagedcode/models.py [hunk 11]
```diff
@@ -135,153 +210,121 @@ class Package(object):
...
+    bug_tracking_url = URLType()
+
+    # strings (such as email, urls, etc)
+    support_contacts = ListType(StringType(), default=[])
+
+    # a URL where the code can be browsed online
+    code_view_url = URLType()
+
+    # one of git, svn, hg, etc
+    vcs_tool = StringType(choices=['git', 'svn', 'hg'])
```
Comment (pombredanne) [inline src/packagedcode/models.py]: add also `cvs`

File: src/packagedcode/models.py [hunk 12]
```diff
@@ -94,14 +103,80 @@
   -- code in source or compiled form or both.
 """
 
-class Package(object):
+
+class EnhancedModel(Model):
+    def __init__(self, deserialize_mapping=None, strict=True, **kwargs):
```
Comment (pombredanne) [inline src/packagedcode/models.py]: What are `deserialize_mapping=None, strict=True` about? 
Also the class should be rather called `BaseModel` IMH

File: src/packagedcode/models.py [hunk 13]
```diff
@@ -94,14 +103,80 @@
...
+        Model.__init__(self, raw_data=kwargs, deserialize_mapping=deserialize_mapping, strict=strict)
+
+
+class Version(EnhancedModel):
```
Comment (pombredanne) [inline src/packagedcode/models.py]: I am a tad confused about Version being a model vs being what used to be called versioning? can you explain how you see this work in practice?

File: src/packagedcode/models.py [hunk 14]
```diff
@@ -94,14 +103,80 @@
...
+    version = StringType()
+
+
+class AssertedLicense(EnhancedModel):
+    """
+    As asserted in a package
+    """
+    license = StringType(default=None)
```
Comment (pombredanne) [inline src/packagedcode/models.py]: Why do these strings have a default to None and most other strings have no defaults? what is the purpose?

File: src/packagedcode/models.py [hunk 15]
```diff
@@ -94,14 +103,80 @@
...
+    version_constraint = StringType()
+
+    # a normalized list of version constraints for this dep.
+    # this is package indepdenent
+    normalized_version_constraints = ListType(StringType(), default=[])
+
+    # is the dep up to date and the latest available?
+    is_latest = BooleanType(default=False)
+
+    def resolve_and_normalize(self):
```
Comment (pombredanne) [inline src/packagedcode/models.py]: What is the purpose of this new method?

File: src/packagedcode/models.py [hunk 16]
```diff
@@ -94,14 +103,80 @@
...
+    id = StringType()
+    # the effective or concrete resolved and used version
+    version = StringType()
+
+    # the potential constraints for this dep.
+    # this is package format specific
+    version_constraint = StringType()
+
+    # a normalized list of version constraints for this dep.
+    # this is package indepdenent
```
Comment (pombredanne) [inline src/packagedcode/models.py]: There is typo in the comment. But I would need to get some explanation on how you see versioning working first

File: src/packagedcode/models.py (hunk 15 above)
Comment (balusarakesh) [inline src/packagedcode/models.py]: it was there in the original Package class `https://github.com/nexB/scancode-toolkit/blob/develop/src/packagedcode/models.py#L421`

File: thirdparty/prod/schematics.ABOUT (hunk 4 above)
Comment (balusarakesh) [inline thirdparty/prod/schematics.ABOUT]: Added `schematics.LICENSE` and copyrigths.

File: src/packagedcode/models.py (hunk 6 above)
Comment (balusarakesh) [inline src/packagedcode/models.py]: fixed the typo

File: src/packagedcode/models.py (hunk 10 above)
Comment (balusarakesh) [inline src/packagedcode/models.py]: removed URLType and using StringType instead

File: src/packagedcode/models.py (hunk 13 above)
Comment (balusarakesh) [inline src/packagedcode/models.py]: renamed class `Version` to `Versioning` to make it more readable

File: src/packagedcode/models.py (hunk 14 above)
Comment (balusarakesh) [inline src/packagedcode/models.py]: removed default values when not required

File: src/packagedcode/models.py (hunk 16 above)
Comment (balusarakesh) [inline src/packagedcode/models.py]: fixed the typo

File: src/packagedcode/models.py (hunk 9 above)
Comment (balusarakesh) [inline src/packagedcode/models.py]: using StringType instead of URLType

File: tests/commoncode/test_fileutils.py [hunk 17]
```diff
@@ -318,14 +318,14 @@ def test_os_walk_with_unicode_path(self):
     def test_fileutils_walk(self):
         test_dir = self.get_test_loc('fileutils/walk')
         base = self.get_test_loc('fileutils')
-        result = [(as_posixpath(t.replace(base, '')), d, f,) for t, d, f in fileutils.walk(test_dir)]
+        result = [(as_posixpath(t.replace(base, '')), d, sorted(f),) for t, d, f in fileutils.walk(test_dir)]
```
Comment (pombredanne) [inline tests/commoncode/test_fileutils.py]: How could a change to these tests be related and part of the package model validation?

File: tests/packagedcode/test_package_models.py [hunk 18]
```diff
@@ -22,23 +22,146 @@
...
     test_data_dir = os.path.join(os.path.dirname(__file__), 'data')
 
     def test_model_creation_and_dump(self):
-        aap = models.AndroidAppPackage()
+        aap = models.AndroidAppPackage(name='someAndroidPAcakge')
         result = aap.as_dict()
-        expected = OrderedDict([('type', 'Android app'),
-                                ('packaging', 'archive'),
-                                ('primary_language', 'Java')])
+        expected = {
```
Comment (pombredanne) [inline tests/packagedcode/test_package_models.py]: We need the as_dict() function to return an OrderedDict with a stable an well defined order. Otherwise the results of a identical scan may not be the same text-wise. We want them to be stable and in a predictable order.
Could you return an OrderedDict again?

File: thirdparty/prod/schematics.LICENSE [hunk 19]
```diff
@@ -0,0 +1,27 @@
+(The BSD License)
```
Comment (pombredanne) [inline thirdparty/prod/schematics.LICENSE]: You probably want to delete this file and rename the .NOTICE to .LICENSE

File: tests/commoncode/test_fileutils.py (hunk 17 above)
Comment (balusarakesh) [inline tests/commoncode/test_fileutils.py]: the tests failed in my windows PC when I removed the `sorted()` function.

Comment (balusarakesh) [inline tests/commoncode/test_fileutils.py]: because the `result` attribute have different values in Ubuntu and Windows : 
windows : 
```
expected = [
            ('/walk', ['d1'], ['f', 'unicode.zip']),
            ('/walk/d1', ['d2'], ['f1']),
            ('/walk/d1/d2', ['d3'], ['f2']),
            ('/walk/d1/d2/d3', [], ['f3'])
        ]
```

Ubuntu : 
```
expected = [
            ('/walk', ['d1'], ['unicode.zip', 'f']),
            ('/walk/d1', ['d2'], ['f1']),
            ('/walk/d1/d2', ['d3'], ['f2']),
            ('/walk/d1/d2/d3', [], ['f3'])
        ]
```
So, I put a `sorted()` which fixed the issue.

File: tests/packagedcode/test_package_models.py (hunk 18 above)
Comment (balusarakesh) [inline tests/packagedcode/test_package_models.py]: the `as_dict()` function returns `OrderedDict`, please check this URL - `https://github.com/nexB/scancode-toolkit/blob/241-new-package-model/src/packagedcode/models.py#L306`

File: tests/commoncode/test_fileutils.py (hunk 17 above)
Comment (pombredanne) [inline tests/commoncode/test_fileutils.py]: ok, this make sense. Can you put that in a completely different pull request?
I would rather not mix this with the package model. They are completely unrelated things.

File: tests/packagedcode/test_package_models.py (hunk 18 above)
Comment (pombredanne) [inline tests/packagedcode/test_package_models.py]: But if you return an ordered dict, why did you change the test to use a bare dict?

File: tests/commoncode/test_fileutils.py (hunk 17 above)
Comment (balusarakesh) [inline tests/commoncode/test_fileutils.py]: Ok

File: tests/packagedcode/test_package_models.py (hunk 18 above)
Comment (pombredanne) [inline tests/packagedcode/test_package_models.py]: Actually, this is unlikely yield a stable ordering:
https://github.com/nexB/scancode-toolkit/blob/829ea55b4d11e4c8ad379a2e081cd7f590b07e81/src/packagedcode/models.py#L306.
If `self.to_primitive().items()` return a bare dict, then you are not really creating an ordereddict with a stable order. You can wrapping a random dict in an ordereddict. Still random in the end.

File: thirdparty/prod/schematics.LICENSE (hunk 19 above)
Comment (balusarakesh) [inline thirdparty/prod/schematics.LICENSE]: fixed it

File: tests/packagedcode/test_package_models.py (hunk 18 above)
Comment (balusarakesh) [inline tests/packagedcode/test_package_models.py]: fixed it - please check - https://github.com/nexB/scancode-toolkit/blob/241-new-package-model/src/packagedcode/models.py#L305

**Comment by pombredanne** [comment]
I am merging this in develop now. Thanks @rakeshbalusa !
In hindsight the use of schematics may not be my best choice, but we can fix that later
